File: pyrobird/pyrobird/edm4hep.py
import logging
import math
from collections import defaultdict

import awkward as ak

from pyrobird.dex import make_dex, PIECE_VERSION

"""
Converts EDM4hep simulation output (ddsim / DD4hep) to Firebird DEX format.

Supported collections:
    vector<edm4hep::SimTrackerHitData> - simulated tracker hits, converted to BoxHit groups
    vector<edm4hep::MCParticleData>    - MC particles, used to build MC-truth trajectories
                                         connecting the sim hits of each particle

SimTrackerHitData branch layout (edm4hep >= 0.99; older versions use 'EDep' instead of 'eDep'
and name the particle relation '_<Collection>_MCParticle' instead of '_<Collection>_particle'):
    'SiBarrelHits': 'vector<edm4hep::SimTrackerHitData>',
    'SiBarrelHits/SiBarrelHits.cellID': 'uint64_t[]',
    'SiBarrelHits/SiBarrelHits.eDep': 'float[]',
    'SiBarrelHits/SiBarrelHits.time': 'float[]',
    'SiBarrelHits/SiBarrelHits.pathLength': 'float[]',
    'SiBarrelHits/SiBarrelHits.quality': 'int32_t[]',
    'SiBarrelHits/SiBarrelHits.position.x': 'double[]',
    'SiBarrelHits/SiBarrelHits.position.y': 'double[]',
    'SiBarrelHits/SiBarrelHits.position.z': 'double[]',
    'SiBarrelHits/SiBarrelHits.momentum.x': 'float[]',
    'SiBarrelHits/SiBarrelHits.momentum.y': 'float[]',
    'SiBarrelHits/SiBarrelHits.momentum.z': 'float[]',
    '_SiBarrelHits_particle': 'vector<podio::ObjectID>',
    '_SiBarrelHits_particle/_SiBarrelHits_particle.index': 'int32_t[]',
    '_SiBarrelHits_particle/_SiBarrelHits_particle.collectionID': 'uint32_t[]',

Sim hits carry no positionError/timeError/edepError, so BoxHit dimensions use a fixed,
user-configurable box size and errors are written as 0.
"""

logger = logging.getLogger(__name__)

# Sim hits have no positionError, so boxes get a fixed visualization size [mm]
DEFAULT_HIT_BOX_SIZE = 2.0

# Cherenkov/PID collections record photon detections attributed to the emitting charged
# particle (optical photons are not stored in MCParticles). Connecting them into that
# particle's trajectory draws zigzags along the photosensor planes, so they are excluded
# from trajectory building by default (they are still converted as BoxHit groups).
DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS = ("DIRCBarHits", "DRICHHits", "PFRICHHits")


def get_sim_hit_branches(tree):
    """Returns names of all vector<edm4hep::SimTrackerHitData> branches in the tree"""
    typenames = tree.typenames(recursive=False, full_paths=True,
                               filter_typename="vector<edm4hep::SimTrackerHitData>")
    return list(typenames.keys())


def detect_file_type(tree):
    """
    Detects the data model of an opened 'events' tree by its branch types.

    Returns 'edm4eic' if reconstructed edm4eic::TrackerHitData collections are present
    (eicrecon files also contain sim hits, reconstruction wins), 'edm4hep' if only
    edm4hep::SimTrackerHitData collections are present.

    Raises
    ------
    ValueError
        If neither collection type is found in the tree.
    """
    if tree.typenames(recursive=False, full_paths=True, filter_typename="vector<edm4eic::TrackerHitData>"):
        return "edm4eic"
    if tree.typenames(recursive=False, full_paths=True, filter_typename="vector<edm4hep::SimTrackerHitData>"):
        return "edm4hep"
    raise ValueError("Cannot detect file type: no 'vector<edm4eic::TrackerHitData>' or "
                     "'vector<edm4hep::SimTrackerHitData>' branches found in 'events' tree")


def _get_field_array(tree, full_branch, entry_start, entry_stop):
    """Reads a jagged field branch and returns it flattened as a python list"""
    return ak.flatten(tree[full_branch].array(entry_start=entry_start, entry_stop=entry_stop)).to_list()


def _find_particle_relation_branch(tree, branch_name):
    """Finds the hit->MCParticle relation branch name, or None if missing.

    edm4hep >= 0.99 names it '_<Collection>_particle', older versions '_<Collection>_MCParticle'.
    """
    keys = set(tree.keys(recursive=False))
    for suffix in ("particle", "MCParticle"):
        candidate = f"_{branch_name}_{suffix}"
        if candidate in keys:
            return candidate
    return None


def _find_edep_field(tree, branch_name):
    """Returns the energy deposit field name: 'eDep' (edm4hep >= 0.99) or 'EDep' (older)"""
    keys = set(tree[branch_name].keys())
    for field in ("eDep", "EDep"):
        if f"{branch_name}.{field}" in keys:
            return field
    return None


def _get_mc_collection_id(tree, mc_branch):
    """
    Looks up the podio collectionID of `mc_branch` from the file's podio_metadata tree.
    Returns None if the metadata is unavailable (e.g. trimmed test files).
    """
    try:
        metadata_tree = tree.file["podio_metadata"]
        info_branch = "events___CollectionTypeInfo"
        ids = metadata_tree[f"{info_branch}/{info_branch}.collectionID"].array().tolist()[0]
        names = metadata_tree[f"{info_branch}/{info_branch}.name"].array().tolist()[0]
        for coll_id, name in zip(ids, names):
            if name == mc_branch:
                return coll_id
    except Exception as ex:
        logger.debug(f"podio_metadata not available for collectionID check: {ex}")
    return None


def sim_tracker_hits_to_box_hits(tree, branch_name, entry_start, entry_stop=None,
                                 box_size=DEFAULT_HIT_BOX_SIZE):
    """Converts vector<edm4hep::SimTrackerHitData> to BoxHit format dictionary.

    Sim hits have no positionError, so all boxes use a fixed `box_size` [mm] cube
    and time/energy errors are written as 0.
    """

    # Read only 1 event if entry_stop is not given
    if entry_stop is None:
        entry_stop = entry_start + 1

    def get_field_array(field):
        return _get_field_array(tree, f"{branch_name}/{branch_name}.{field}", entry_start, entry_stop)

    pos_x = get_field_array("position.x")   # 'double[]'
    pos_y = get_field_array("position.y")   # 'double[]'
    pos_z = get_field_array("position.z")   # 'double[]'
    time = get_field_array("time")          # 'float[]'

    edep_field = _find_edep_field(tree, branch_name)
    if edep_field:
        edep = get_field_array(edep_field)
    else:
        logger.warning(f"No eDep/EDep field found in '{branch_name}', writing 0 energy deposit")
        edep = [0.0] * len(pos_x)

    # Columnar piece: parallel arrays, hit id == array index; pos is flat xyz triplets.
    # Sim data carries no errors, so the error columns are omitted entirely.
    count = len(pos_x)
    pos = []
    for i in range(count):
        pos.extend((pos_x[i], pos_y[i], pos_z[i]))

    piece = {
        "name": branch_name,
        "type": "BoxHit",
        "version": PIECE_VERSION,
        "origin": {"type": "edm4hep::SimTrackerHitData", "name": branch_name},
        "count": count,
        "columns": {
            "pos": pos,
            "dim": [box_size] * (3 * count),
            "time": time,
            "edep": edep,
        },
    }
    return piece


def sim_hits_to_trajectories(tree, hit_branch_names, entry_start, entry_stop=None,
                             mc_branch="MCParticles", min_hits=2,
                             prepend_vertex=True, append_endpoint=False):
    """
    Builds MC-truth trajectories connecting edm4hep::SimTrackerHitData hits.

    Hits from all `hit_branch_names` collections are pooled, grouped by their associated
    MCParticle and sorted by time. Each particle with at least `min_hits` hits becomes one
    trajectory in a single 'PointTrajectory' group.

    Parameters
    ----------
    tree - uproot 'events' tree
    hit_branch_names - SimTrackerHitData collection names to pool hits from
    entry_start - first entry to process
    entry_stop - one past last entry (defaults to entry_start + 1, i.e. a single event)
    mc_branch - name of the MCParticle collection the hit relations point to
    min_hits - particles with fewer hits are skipped (their hits stay visible as boxes)
    prepend_vertex - add MCParticles.vertex (with the particle creation time) as point 0
    append_endpoint - add MCParticles.endpoint as the last point; edm4hep stores no endpoint
                      time, so the point reuses the last hit's time
    """
    if entry_stop is None:
        entry_stop = entry_start + 1

    # Columnar piece: track parameters are per-name columns (trajectory id == index),
    # the ragged per-trajectory point lists stay nested under "points"
    result = {
        "name": "McHitTrajectories",
        "type": "PointTrajectory",
        "version": PIECE_VERSION,
        "origin": {"type": ["edm4hep::SimTrackerHitData", "edm4hep::MCParticleData"],
                   "collections": list(hit_branch_names)},
        "count": 0,
        "columns": {},
        "pointColumns": ["x", "y", "z", "t", "dx", "dy", "dz", "dt"],
        "points": []
    }

    if mc_branch not in tree.keys(recursive=False):
        logger.warning(f"'{mc_branch}' collection not found, no trajectories are built")
        return result

    def get_mc_field(field):
        return _get_field_array(tree, f"{mc_branch}/{mc_branch}.{field}", entry_start, entry_stop)

    mc_pdg = get_mc_field("PDG")
    mc_charge = get_mc_field("charge")
    mc_time = get_mc_field("time")
    mc_mom_x = get_mc_field("momentum.x")
    mc_mom_y = get_mc_field("momentum.y")
    mc_mom_z = get_mc_field("momentum.z")
    mc_vtx_x = get_mc_field("vertex.x")
    mc_vtx_y = get_mc_field("vertex.y")
    mc_vtx_z = get_mc_field("vertex.z")
    mc_end_x = get_mc_field("endpoint.x") if append_endpoint else []
    mc_end_y = get_mc_field("endpoint.y") if append_endpoint else []
    mc_end_z = get_mc_field("endpoint.z") if append_endpoint else []

    # eicrecon files carry a second MCParticle collection (MCParticlesHeadOnFrameNoBeamFX);
    # all known sim-hit relations point to 'MCParticles', but sanity-check via collectionID
    # when podio metadata is available
    mc_collection_id = _get_mc_collection_id(tree, mc_branch)

    # Pool hits from all collections grouped by MCParticle index
    # Each hit is a (time, x, y, z) tuple
    hits_by_particle = defaultdict(list)
    for branch_name in hit_branch_names:
        relation_branch = _find_particle_relation_branch(tree, branch_name)
        if relation_branch is None:
            logger.warning(f"No MCParticle relation branch found for '{branch_name}', "
                           f"its hits are skipped in trajectory building")
            continue

        def get_field_array(field):
            return _get_field_array(tree, f"{branch_name}/{branch_name}.{field}", entry_start, entry_stop)

        pos_x = get_field_array("position.x")
        pos_y = get_field_array("position.y")
        pos_z = get_field_array("position.z")
        time = get_field_array("time")
        mc_index = _get_field_array(tree, f"{relation_branch}/{relation_branch}.index", entry_start, entry_stop)
        coll_ids = _get_field_array(tree, f"{relation_branch}/{relation_branch}.collectionID", entry_start, entry_stop)

        for i in range(len(pos_x)):
            index = mc_index[i]
            if index < 0 or index >= len(mc_pdg):
                continue  # unfilled relation
            if mc_collection_id is not None and coll_ids[i] != mc_collection_id:
                logger.warning(f"Hit in '{branch_name}' points to an unexpected MCParticle "
                               f"collection (collectionID={coll_ids[i]}), skipping it")
                continue
            hits_by_particle[index].append((time[i], pos_x[i], pos_y[i], pos_z[i]))

    columns = {name: [] for name in ("pdg", "charge", "px", "py", "pz", "p", "mc_index")}
    all_points = []
    for index in sorted(hits_by_particle.keys()):
        particle_hits = hits_by_particle[index]
        if len(particle_hits) < min_hits:
            continue
        particle_hits.sort(key=lambda hit: hit[0])

        # pointColumns => [x, y, z, t, dx, dy, dz, dt]; sim data has no errors
        points = []
        if prepend_vertex:
            points.append([mc_vtx_x[index], mc_vtx_y[index], mc_vtx_z[index], mc_time[index], 0, 0, 0, 0])
        for hit_time, x, y, z in particle_hits:
            points.append([x, y, z, hit_time, 0, 0, 0, 0])
        if append_endpoint:
            # edm4hep stores no endpoint time, reuse the last hit time to keep time animation monotonic
            last_time = particle_hits[-1][0]
            points.append([mc_end_x[index], mc_end_y[index], mc_end_z[index], last_time, 0, 0, 0, 0])

        momentum = math.sqrt(mc_mom_x[index] ** 2 + mc_mom_y[index] ** 2 + mc_mom_z[index] ** 2)
        columns["pdg"].append(mc_pdg[index])
        columns["charge"].append(mc_charge[index])
        columns["px"].append(mc_mom_x[index])
        columns["py"].append(mc_mom_y[index])
        columns["pz"].append(mc_mom_z[index])
        columns["p"].append(momentum)
        columns["mc_index"].append(index)
        all_points.append(points)

    result["count"] = len(all_points)
    result["columns"] = columns if all_points else {}
    result["points"] = all_points
    return result


def edm4hep_entry_to_dict(tree, entry_index, custom_name=None, collections=None,
                          box_size=DEFAULT_HIT_BOX_SIZE,
                          prepend_vertex=True, append_endpoint=False,
                          trajectory_excluded_collections=DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS):
    # the result of this function
    components = []

    if not collections:
        collections = [
            "tracker_hits",
            "mc_trajectories"
        ]

    hit_branches = get_sim_hit_branches(tree)

    # Hits:
    if "tracker_hits" in collections:
        for branch_name in hit_branches:
            piece = sim_tracker_hits_to_box_hits(tree, branch_name, entry_index, box_size=box_size)
            # Collections empty in this event are skipped
            if piece["count"] > 0:
                components.append(piece)

    # MC-truth trajectories connecting the hits
    if "mc_trajectories" in collections:
        trajectory_branches = [b for b in hit_branches if b not in trajectory_excluded_collections]
        trajectory_piece = sim_hits_to_trajectories(tree, trajectory_branches, entry_index,
                                                    prepend_vertex=prepend_vertex,
                                                    append_endpoint=append_endpoint)
        if trajectory_piece["count"] > 0:
            components.append(trajectory_piece)

    entry = {
        "id": custom_name if custom_name else entry_index,
        "pieces": components
    }

    return entry


def edm4hep_to_dex_dict(tree, event_ids, origin_info=None, collections=None,
                        box_size=DEFAULT_HIT_BOX_SIZE,
                        prepend_vertex=True, append_endpoint=False,
                        trajectory_excluded_collections=DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS):
    event_data = []

    if isinstance(event_ids, int):
        event_ids = [event_ids]

    for entry_id in event_ids:
        event_data.append(edm4hep_entry_to_dict(
            tree, entry_id, custom_name=None, collections=collections,
            box_size=box_size, prepend_vertex=prepend_vertex, append_endpoint=append_endpoint,
            trajectory_excluded_collections=trajectory_excluded_collections))

    return make_dex(event_data, origin_info)

File: packages/root2dex/src/edm4hep.ts
/**
 * EDM4hep (ddsim / DD4hep simulation output) -> Firebird DEX.
 *
 * Mirrors pyrobird's `pyrobird/edm4hep.py` piece for piece.
 *
 * Collections converted:
 *   vector<edm4hep::SimTrackerHitData> -> one BoxHit piece per collection
 *   vector<edm4hep::MCParticleData>    -> MC-truth trajectories connecting the
 *                                         sim hits of each particle
 *
 * Sim hits carry no positionError/timeError/edepError, so BoxHit dimensions use
 * a fixed configurable box size and the error columns are omitted entirely.
 */

import { PIECE_VERSION, type DexEvent, type DexPiece } from './dex';
import { numbers, type ColumnBag, type PodioEventFile } from './podio-file';
import {
  mcParticlesBranchNames,
  mcParticlesToTrajectories,
  type McParticlesOptions,
} from './mc-particles';

export const EDM4HEP_SIM_HIT_TYPE = 'vector<edm4hep::SimTrackerHitData>';

/** Sim hits have no positionError, so boxes get a fixed visualization size [mm]. */
export const DEFAULT_HIT_BOX_SIZE = 2.0;

/**
 * Cherenkov/PID collections record photon detections attributed to the emitting
 * charged particle (optical photons are not stored in MCParticles). Connecting
 * them into that particle's trajectory draws zigzags along the photosensor
 * planes, so they are excluded from trajectory building by default (they are
 * still converted as BoxHit pieces).
 */
export const DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS = ['DIRCBarHits', 'DRICHHits', 'PFRICHHits'];

/** Which collection groups to convert. Defaults to all of them. */
export type Edm4hepCollection = 'tracker_hits' | 'mc_trajectories' | 'mc_particles';
export const EDM4HEP_DEFAULT_COLLECTIONS: Edm4hepCollection[] = [
  'tracker_hits',
  'mc_trajectories',
  'mc_particles',
];

export interface Edm4hepOptions extends McParticlesOptions {
  collections?: Edm4hepCollection[];
  /** Box size in mm for sim hits, which carry no position error. */
  boxSize?: number;
  /** Prepend the MCParticle vertex (with its creation time) as trajectory point 0. */
  prependVertex?: boolean;
  /** Append the MCParticle endpoint as the last trajectory point. */
  appendEndpoint?: boolean;
  /** Particles with fewer hits are skipped; their hits stay visible as boxes. */
  minHits?: number;
  /** Collection name the sim-hit relations point to. */
  mcBranch?: string;
  trajectoryExcludedCollections?: string[];
  onWarning?: (message: string) => void;
}

const MC_FIELDS_ALWAYS = [
  'PDG',
  'charge',
  'time',
  'momentum.x',
  'momentum.y',
  'momentum.z',
  'vertex.x',
  'vertex.y',
  'vertex.z',
];
const MC_FIELDS_ENDPOINT = ['endpoint.x', 'endpoint.y', 'endpoint.z'];

/**
 * Finds the hit -> MCParticle relation branch, or null when missing.
 * edm4hep >= 0.99 names it '_<Collection>_particle', older versions
 * '_<Collection>_MCParticle'.
 */
export function findParticleRelationBranch(file: PodioEventFile, name: string): string | null {
  for (const suffix of ['particle', 'MCParticle']) {
    const candidate = `_${name}_${suffix}`;
    if (file.hasBranch(candidate)) return candidate;
  }
  return null;
}

/** Returns the energy deposit field name: 'eDep' (edm4hep >= 0.99) or 'EDep' (older). */
export function findEdepField(file: PodioEventFile, name: string): string | null {
  for (const field of ['eDep', 'EDep']) {
    if (file.hasBranch(`${name}.${field}`)) return field;
  }
  return null;
}

/**
 * Every branch an edm4hep entry conversion reads. Collected up front so one
 * `readEntry` pass covers the whole event - see PodioEventFile.
 */
export function edm4hepBranchNames(file: PodioEventFile, options: Edm4hepOptions = {}): string[] {
  const collections = options.collections?.length
    ? options.collections
    : EDM4HEP_DEFAULT_COLLECTIONS;
  const excluded =
    options.trajectoryExcludedCollections ?? DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS;
  const mcBranch = options.mcBranch ?? 'MCParticles';
  const hitBranches = file.collectionsOfType(EDM4HEP_SIM_HIT_TYPE);
  const names: string[] = [];

  if (collections.includes('tracker_hits')) {
    for (const name of hitBranches) {
      names.push(`${name}.position.x`, `${name}.position.y`, `${name}.position.z`, `${name}.time`);
      const edep = findEdepField(file, name);
      if (edep) names.push(`${name}.${edep}`);
    }
  }

  if (collections.includes('mc_trajectories')) {
    for (const field of MC_FIELDS_ALWAYS) names.push(`${mcBranch}.${field}`);
    if (options.appendEndpoint) {
      for (const field of MC_FIELDS_ENDPOINT) names.push(`${mcBranch}.${field}`);
    }
    for (const name of hitBranches) {
      if (excluded.includes(name)) continue;
      names.push(`${name}.position.x`, `${name}.position.y`, `${name}.position.z`, `${name}.time`);
      const relation = findParticleRelationBranch(file, name);
      if (relation) names.push(`${relation}.index`, `${relation}.collectionID`);
    }
  }

  if (collections.includes('mc_particles')) {
    names.push(...mcParticlesBranchNames(options));
  }

  return names;
}

/** Converts one vector<edm4hep::SimTrackerHitData> collection to a BoxHit piece. */
export function simTrackerHitsToBoxHits(
  file: PodioEventFile,
  bag: ColumnBag,
  name: string,
  boxSize = DEFAULT_HIT_BOX_SIZE,
  onWarning?: (message: string) => void,
): DexPiece {
  const posX = numbers(bag, `${name}.position.x`);
  const posY = numbers(bag, `${name}.position.y`);
  const posZ = numbers(bag, `${name}.position.z`);
  const count = posX.length;

  const edepField = findEdepField(file, name);
  let edep: number[];
  if (edepField) {
    edep = numbers(bag, `${name}.${edepField}`);
  } else {
    onWarning?.(`No eDep/EDep field found in '${name}', writing 0 energy deposit`);
    edep = new Array<number>(count).fill(0);
  }

  const pos: number[] = [];
  for (let i = 0; i < count; i++) pos.push(posX[i], posY[i], posZ[i]);

  return {
    name,
    type: 'BoxHit',
    version: PIECE_VERSION,
    origin: { type: 'edm4hep::SimTrackerHitData', name },
    count,
    columns: {
      // Sim data carries no errors, so the error columns are omitted entirely
      pos,
      dim: new Array<number>(3 * count).fill(boxSize),
      time: numbers(bag, `${name}.time`),
      edep,
    },
  };
}

/**
 * Builds MC-truth trajectories connecting SimTrackerHit hits.
 *
 * Hits from all `hitBranchNames` collections are pooled, grouped by their
 * associated MCParticle and sorted by time. Each particle with at least
 * `minHits` hits becomes one trajectory in a single PointTrajectory piece.
 */
export function simHitsToTrajectories(
  file: PodioEventFile,
  bag: ColumnBag,
  hitBranchNames: string[],
  options: Edm4hepOptions & { mcCollectionId?: number | null } = {},
): DexPiece {
  const mcBranch = options.mcBranch ?? 'MCParticles';
  const minHits = options.minHits ?? 2;
  const prependVertex = options.prependVertex ?? true;
  const appendEndpoint = options.appendEndpoint ?? false;
  const onWarning = options.onWarning;

  const piece: DexPiece = {
    name: 'McHitTrajectories',
    type: 'PointTrajectory',
    version: PIECE_VERSION,
    origin: {
      type: ['edm4hep::SimTrackerHitData', 'edm4hep::MCParticleData'],
      collections: [...hitBranchNames],
    },
    count: 0,
    columns: {},
    pointColumns: ['x', 'y', 'z', 't', 'dx', 'dy', 'dz', 'dt'],
    points: [],
  };

  if (!file.hasBranch(mcBranch)) {
    onWarning?.(`'${mcBranch}' collection not found, no trajectories are built`);
    return piece;
  }

  const mc = (field: string) => numbers(bag, `${mcBranch}.${field}`);
  const mcPdg = mc('PDG');
  const mcCharge = mc('charge');
  const mcTime = mc('time');
  const mcMomX = mc('momentum.x');
  const mcMomY = mc('momentum.y');
  const mcMomZ = mc('momentum.z');
  const mcVtxX = mc('vertex.x');
  const mcVtxY = mc('vertex.y');
  const mcVtxZ = mc('vertex.z');
  const mcEndX = appendEndpoint ? mc('endpoint.x') : [];
  const mcEndY = appendEndpoint ? mc('endpoint.y') : [];
  const mcEndZ = appendEndpoint ? mc('endpoint.z') : [];

  // eicrecon files carry a second MCParticle collection
  // (MCParticlesHeadOnFrameNoBeamFX); all known sim-hit relations point to
  // 'MCParticles', but sanity-check via collectionID when podio metadata is
  // available
  const mcCollectionId = options.mcCollectionId ?? null;

  // Pool hits from all collections grouped by MCParticle index.
  // Each hit is [time, x, y, z].
  const hitsByParticle = new Map<number, number[][]>();
  for (const name of hitBranchNames) {
    const relation = findParticleRelationBranch(file, name);
    if (relation === null) {
      onWarning?.(
        `No MCParticle relation branch found for '${name}', ` +
          `its hits are skipped in trajectory building`,
      );
      continue;
    }
    const posX = numbers(bag, `${name}.position.x`);
    const posY = numbers(bag, `${name}.position.y`);
    const posZ = numbers(bag, `${name}.position.z`);
    const time = numbers(bag, `${name}.time`);
    const mcIndex = numbers(bag, `${relation}.index`);
    const collectionIds = numbers(bag, `${relation}.collectionID`);

    for (let i = 0; i < posX.length; i++) {
      const index = mcIndex[i];
      if (index < 0 || index >= mcPdg.length) continue; // unfilled relation
      if (mcCollectionId !== null && collectionIds[i] !== mcCollectionId) {
        onWarning?.(
          `Hit in '${name}' points to an unexpected MCParticle collection ` +
            `(collectionID=${collectionIds[i]}), skipping it`,
        );
        continue;
      }
      let hits = hitsByParticle.get(index);
      if (!hits) {
        hits = [];
        hitsByParticle.set(index, hits);
      }
      hits.push([time[i], posX[i], posY[i], posZ[i]]);
    }
  }

  const columnNames = ['pdg', 'charge', 'px', 'py', 'pz', 'p', 'mc_index'] as const;
  const columns: Record<string, number[]> = {};
  for (const columnName of columnNames) columns[columnName] = [];
  const allPoints: number[][][] = [];

  for (const index of [...hitsByParticle.keys()].sort((a, b) => a - b)) {
    const particleHits = hitsByParticle.get(index)!;
    if (particleHits.length < minHits) continue;
    particleHits.sort((a, b) => a[0] - b[0]);

    // pointColumns => [x, y, z, t, dx, dy, dz, dt]; sim data has no errors
    const points: number[][] = [];
    if (prependVertex) {
      points.push([mcVtxX[index], mcVtxY[index], mcVtxZ[index], mcTime[index], 0, 0, 0, 0]);
    }
    for (const [hitTime, x, y, z] of particleHits) {
      points.push([x, y, z, hitTime, 0, 0, 0, 0]);
    }
    if (appendEndpoint) {
      // edm4hep stores no endpoint time, reuse the last hit time to keep time
      // animation monotonic
      const lastTime = particleHits[particleHits.length - 1][0];
      points.push([mcEndX[index], mcEndY[index], mcEndZ[index], lastTime, 0, 0, 0, 0]);
    }

    const momentum = Math.sqrt(
      mcMomX[index] ** 2 + mcMomY[index] ** 2 + mcMomZ[index] ** 2,
    );
    columns['pdg'].push(mcPdg[index]);
    columns['charge'].push(mcCharge[index]);
    columns['px'].push(mcMomX[index]);
    columns['py'].push(mcMomY[index]);
    columns['pz'].push(mcMomZ[index]);
    columns['p'].push(momentum);
    columns['mc_index'].push(index);
    allPoints.push(points);
  }

  piece.count = allPoints.length;
  piece.columns = allPoints.length ? columns : {};
  piece.points = allPoints;
  return piece;
}

/** Converts one entry of an edm4hep 'events' tree to a DEX event object. */
export async function edm4hepEntryToDex(
  file: PodioEventFile,
  entry: number,
  options: Edm4hepOptions & { mcCollectionId?: number | null } = {},
): Promise<DexEvent> {
  const collections = options.collections?.length
    ? options.collections
    : EDM4HEP_DEFAULT_COLLECTIONS;
  const excluded =
    options.trajectoryExcludedCollections ?? DEFAULT_TRAJECTORY_EXCLUDED_COLLECTIONS;
  const boxSize = options.boxSize ?? DEFAULT_HIT_BOX_SIZE;

  const bag = await file.readEntry(edm4hepBranchNames(file, options), entry);
  const hitBranches = file.collectionsOfType(EDM4HEP_SIM_HIT_TYPE);
  const pieces: DexPiece[] = [];

  if (collections.includes('tracker_hits')) {
    for (const name of hitBranches) {
      const piece = simTrackerHitsToBoxHits(file, bag, name, boxSize, options.onWarning);
      // Collections empty in this event are skipped
      if (piece.count > 0) pieces.push(piece);
    }
  }

  if (collections.includes('mc_trajectories')) {
    const trajectoryBranches = hitBranches.filter(name => !excluded.includes(name));
    const piece = simHitsToTrajectories(file, bag, trajectoryBranches, options);
    if (piece.count > 0) pieces.push(piece);
  }

  if (collections.includes('mc_particles')) {
    const piece = mcParticlesToTrajectories(file, bag, options);
    if (piece.count > 0) pieces.push(piece);
  }

  return { id: entry, pieces };
}
